### PR TITLE
[NEW] add delete-own-message permission

### LINKED ISSUE #5434: Can delete own messages even without the delete message permission
If you remove the delete-message and edit-message permission, the user can still delete and edit his own message. Now its impossible to block the user from editing or deleting his messages without blocking everybody from doing that. 
My suggestion is to create a new permission for delete-own-messages and edit-own-messages.

### PR DESCRIPTION
**Pull request by ggazzo**
Closes #5434

### CODE REVIEW

**Comment by rodrigok** [comment]
@ggazzo need to improve the title!

**Comment by ggazzo** [comment]
@rodrigok thinking more about, I dont think this pr is good enough, I'm using return `true`, `false` AND an exception, and I'm not sure if we should use this behavior  with a function `canSomething`...

**Comment by cchbr** [comment]
Will this be included in a release any time soon? It is greatly needed in my instance of Rocket chat.

**Comment by reetp** [comment]
> Will this be included in a release any time soon? It is greatly needed in my instance of Rocket chat.

It's Open Source. So it will be included when it is ready... If you look it is in the 2.3.0 milestone. But that may slip depending on the developers priorities - what is important to you may not be so important to them....

Of course you can pay for support. Or your pull request with a fix will be welcomed. 

If not I'm afraid you will just have to be patient.

**Comment by ggazzo** [comment]
@cchbr sure, we are planning to deliver in to the next release (2.3.0 or 3.0.0,  November 27th), if you are running your own build, you can make a cherry pick :)

File: app/authorization/server/functions/canDeleteMessage.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+		return false;
+	}
+	const blockDeleteInMinutes = await getValue('Message_AllowDeleting_BlockDeleteInMinutes');
+
+	if (!blockDeleteInMinutes) {
+		return true;
+	}
+
+	const currentTsDiff = diff(ts);
+	return currentTsDiff <= blockDeleteInMinutes;
```
Comment (tassoevan) [inline app/authorization/server/functions/canDeleteMessage.js]:
```suggestion
	const timeElapsedForMessage = elapsedTime(ts);
	return timeElapsedForMessage <= blockDeleteInMinutes;
```

File: app/utils/client/lib/canDeleteMessage.js [hunk 2]
```diff
@@ -17,9 +17,10 @@ export const canDeleteMessage = ({ rid, ts, uid }) => {
 		return false;
 	}
 
-	const hasPermission = hasAtLeastOnePermission('delete-message', rid);
-	const deleteOwn = uid === userId;
-	if (!hasPermission && !deleteOwn) {
+	const allowed = hasPermission('delete-message', rid);
+
+	const deleteOwn = allowed || (uid === userId && hasPermission('delete-own-message'));
```
Comment (sampaiodiego) [inline app/utils/client/lib/canDeleteMessage.js]: either remove the `allowed` here or remove it from the `if` below
```suggestion
	const deleteOwn = uid === userId && hasPermission('delete-own-message');
```

File: app/ui-utils/client/lib/messageContext.js [hunk 3]
```diff
@@ -30,6 +30,7 @@ export function messageContext({ rid } = Template.instance()) {
 			showreply: true,
 			showReplyButton: true,
 			hasPermissionDeleteMessage: hasPermission('delete-message', rid),
+			hasPermissionDeleteOwnMessage: hasPermission('delete-own-message'),
```
Comment (sampaiodiego) [inline app/ui-utils/client/lib/messageContext.js]: is this being used somewhere?

Comment (ggazzo) [reply app/ui-utils/client/lib/messageContext.js]: not yet, but `hasPermissionDeleteMessage` is not being used too, I just passing the context, and maybe after I could improve some stuffs.

File: app/utils/client/lib/canDeleteMessage.js (hunk 2 above)
Comment (ggazzo) [reply app/utils/client/lib/canDeleteMessage.js]: I'm skipping one hasPermission, if you are allowed any messages you are already allowed to edit your own messages... we save 3~4 finds on the database...

**Pull request merged**